core/src/avm2/method.rs

```diff
@@ -301,21 +302,87 @@ impl<'gc> fmt::Debug for NativeMethod<'gc> {
     }
 }
 
+#[derive(Clone, Collect, Debug, Copy, PartialEq, Eq)]
+#[collect(no_drop)]
+pub enum MethodPosition {
+    ClassTrait,
+    InstanceTrait,
+}
+
+#[derive(Clone, Collect, Debug, Copy, PartialEq, Eq)]
+#[collect(no_drop)]
+pub enum MethodKind {
+    Initializer,
+    Setter,
+    Getter,
+    Regular,
+}
+
+#[derive(Clone, Collect, Debug, Copy)]
+#[collect(no_drop)]
+pub struct MethodMetadata<'gc> {
+    name: AvmString<'gc>,
+    position: MethodPosition,
+    kind: MethodKind,
+}
+
+impl<'gc> MethodMetadata<'gc> {
+    pub fn from_trait(my_trait: &Trait<'gc>, position: MethodPosition) -> Self {
+        Self {
+            name: my_trait.name().local_name(),
+            position,
+            kind: match my_trait.kind() {
+                TraitKind::Getter { .. } => MethodKind::Getter,
+                TraitKind::Setter { .. } => MethodKind::Setter,
+                _ => MethodKind::Regular,
+            },
+        }
+    }
+
+    pub fn new_class_init() -> Self {
+        Self {
+            name: "cinit".into(),
+            position: MethodPosition::ClassTrait,
+            kind: MethodKind::Initializer,
+        }
+    }
+
+    pub fn new_instance_init() -> Self {
+        Self {
+            name: AvmString::default(),
+            position: MethodPosition::InstanceTrait,
+            kind: MethodKind::Initializer,
+        }
+    }
+
+    pub fn name(&self) -> AvmString<'gc> {
+        self.name
+    }
+
+    pub fn position(&self) -> MethodPosition {
+        self.position
+    }
+
+    pub fn kind(&self) -> MethodKind {
+        self.kind
+    }
+}
+
 /// An uninstantiated method that can either be natively implemented or sourced
 /// from an ABC file.
 #[derive(Clone, Collect, Debug)]
 #[collect(no_drop)]
 pub enum Method<'gc> {
     /// A native method.
-    Native(Gc<'gc, NativeMethod<'gc>>),
+    Native(Gc<'gc, NativeMethod<'gc>>, Option<MethodMetadata<'gc>>),
```

Comment:
Hmm, this can be a possible footgun if someone copies the Method and expects all of it (metadata) to still be shared...
To be sure, is the `Clone` derive actually used? If not, maybe let's remove it for safety.

Reply:
The clone derive is used, so I would have to put the metadata in a GcCell, but I feel like this is unnecessarily expensive. I think the metadata not being shared is not a bad thing, since the metadata is automatically set correctly when they are added to a class. It might be worth adding a comment to warn people of this behavior though